[pytorch][logging] add empty wait counter implementation

jamesperng opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #128605
* __->__ #128466



cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

Differential Revision: [D58441466](https://our.internmc.facebook.com/intern/diff/D58441466)

jamesperng committed:
[pytorch][logging] add empty wait counter implementation

[ghstack-poisoned]

jamesperng commented:
@jamesperng has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58441466).

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -582,6 +583,7 @@ bool ProcessGroupNCCL::WorkNCCL::finishedGPUExecutionInternal() const {
 
 bool ProcessGroupNCCL::WorkNCCL::checkTimeout(
     std::optional<std::chrono::milliseconds> timeout) {
+  SCOPED_WAIT_COUNTER_US("pytorch.logging.wait_counter.ProcessGroupNCCL::WorkNCCL::checkTimeout");
```
would it be convention that all counters have the same prefix "pytorch.logging.wait_counter"?  If so maybe its worth prepending that inside the counter impl and then users can just write the suffix "ProcessGroup..."

jamesperng replied:
I think that for these wait counters, we would have the same prefix pytorch.wait_counter (i took out the logging).

jamesperng replied:
the downside of having the prefix put in implicitly, is that it might make the counters less searchable in the code?

jamesperng committed:
Update on "[pytorch][logging] add empty wait counter implementation"




cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

Differential Revision: [D58441466](https://our.internmc.facebook.com/intern/diff/D58441466)

[ghstack-poisoned]

jamesperng committed:
Update on "[pytorch][logging] add empty wait counter implementation"




cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

Differential Revision: [D58441466](https://our.internmc.facebook.com/intern/diff/D58441466)

[ghstack-poisoned]

jamesperng committed:
Update on "[pytorch][logging] add empty wait counter implementation"




cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

Differential Revision: [D58441466](https://our.internmc.facebook.com/intern/diff/D58441466)

[ghstack-poisoned]

jamesperng commented:
/easycla

jamesperng committed:
Update on "[pytorch][logging] add empty wait counter implementation"




cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

Differential Revision: [D58441466](https://our.internmc.facebook.com/intern/diff/D58441466)

[ghstack-poisoned]

jamesperng committed:
Update on "[pytorch][logging] add empty wait counter implementation"




cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

Differential Revision: [D58441466](https://our.internmc.facebook.com/intern/diff/D58441466)

[ghstack-poisoned]

jamesperng committed:
Update on "[pytorch][logging] add empty wait counter implementation"




cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

Differential Revision: [D58441466](https://our.internmc.facebook.com/intern/diff/D58441466)

[ghstack-poisoned]
